src/coreComponents/finiteVolume/SurfaceElementStencil.hpp

```diff
@@ -559,19 +652,99 @@ SurfaceElementStencilWrapper::
       real64 const t0 = mult0 * m_weights[iconn][0] * coefficient[er0][esr0][ei0][0][0];
       real64 const t1 = mult1 * m_weights[iconn][1] * coefficient[er1][esr1][ei1][0][0];
 
-      real64 const harmonicWeight   = t0*t1 / sumOfTrans;
-      real64 const arithmeticWeight = 0.25 * (t0+t1);
-
-      real64 const value = m_meanPermCoefficient * harmonicWeight + (1 - m_meanPermCoefficient) * arithmeticWeight;
 
-      weight[connectionIndex][0] = value;
-      weight[connectionIndex][1] = -value;
 
       connectionIndex++;
     }
   }
 }
 
+GEOS_HOST_DEVICE
+inline void
+SurfaceElementStencilWrapper::averageWeights( const geos::localIndex iconn,
+                                              const geos::localIndex connexionIndex,
+                                              const geos::real64 (& halfWeight)[2],
+                                              const geos::real64 (& dHalfWeight_dVar)[2],
+                                              geos::real64 (& weight)[maxNumConnections][2],
+                                              geos::real64 (& dWeight_dVar)[maxNumConnections][2] ) const
+{
+  real64 sumOfTrans = 0.0;
+  for( localIndex k=0; k<numPointsInFlux( iconn ); ++k )
+  {
+    //TODO check that
+    sumOfTrans += halfWeight[k];
+  }
+
+  real64 const arithmeticWeight = 0.25 * (halfWeight[0]+halfWeight[1]);
+  real64 const harmonicWeight   = halfWeight[0]*halfWeight[1] / sumOfTrans;
+  real64 const value = m_meanPermCoefficient * harmonicWeight + (1 - m_meanPermCoefficient) * arithmeticWeight;
+
+  weight[connexionIndex][0] = value;
+  weight[connexionIndex][1] = -value;
+
+  real64 dHarmonic[2];
+  dHarmonic[0] = ( dHalfWeight_dVar[0] * halfWeight[1] * sumOfTrans - dHalfWeight_dVar[0] * halfWeight[0] * halfWeight[1] ) / ( sumOfTrans * sumOfTrans );
+  dHarmonic[1] = ( dHalfWeight_dVar[1] * halfWeight[0] * sumOfTrans - dHalfWeight_dVar[1] * halfWeight[0] * halfWeight[1] ) / ( sumOfTrans * sumOfTrans );
+
+  real64 dArithmetic[2];
+  dArithmetic[0] = 0.25 * dHalfWeight_dVar[0];
+  dArithmetic[1] = 0.25 * dHalfWeight_dVar[1];
+
+  dWeight_dVar[connexionIndex][0] =   ( m_meanPermCoefficient * dHarmonic[0] + (1 - m_meanPermCoefficient) * dArithmetic[0] );
+  dWeight_dVar[connexionIndex][1] = -( m_meanPermCoefficient * dHarmonic[1] + (1 - m_meanPermCoefficient) * dArithmetic[1] );
+
+
+}
+
+GEOS_HOST_DEVICE
+inline void
+SurfaceElementStencilWrapper::averageWeights( const localIndex iconn,
```

Comment:
I suspect this method is also the same for all stencils.

Reply:
Agree on factoring up possible duplicate.